src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,393 @@
+        goto error;
+    }
+    memcpy(in_data->buffer, in, inl);
+    in_data->size = inl;
+
+    /* Get mode value */
+    mode = tpm2_get_cipher_mode(ctx, tpm2Data);
+
+    /* Get IV and Enc*/
+    iv_in.size = EVP_CIPHER_CTX_iv_length(ctx);
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+    memcpy(iv_in.buffer, ctx->iv, iv_in.size);
+    // Note: Openssl (encrypt:1) != TSS (encrypt:0)
+    enc = !(ctx->encrypt);
+#else
+    memcpy(iv_in.buffer, EVP_CIPHER_CTX_iv(ctx), iv_in.size);
+    enc = !EVP_CIPHER_CTX_encrypting(ctx);
+#endif
+
+    /* Trying to encrypt */
+    ret = Esys_EncryptDecrypt2( eactx.ectx,
+                                keyHandle,
+                                ESYS_TR_PASSWORD,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                in_data,
+                                enc,
+                                mode,
+                                &iv_in,
+                                &out_data,
+                                &iv_out );
+    if (ret == TPM2_RC_COMMAND_CODE) {
+        DBG("Esys_EncryptDecrypt2 : FAILED\n");
+        ret = Esys_EncryptDecrypt( eactx.ectx,
+                                   keyHandle,
+                                   ESYS_TR_PASSWORD,
+                                   ESYS_TR_NONE,
+                                   ESYS_TR_NONE,
+                                   enc,
+                                   mode,
+                                   &iv_in,
+                                   in_data,
+                                   &out_data,
+                                   &iv_out );
+        if(ret == TPM2_RC_SUCCESS) {
+            DBG("Esys_EncryptDecrypt  : SUCCESS\n");
+        }
+        else {
+            DBG("Esys_EncryptDecrypt  : FAILED\n");
+        }
+    }
+    ERRchktss(tpm2_do_cipher, ret, goto error);
+
+    /* Copy out_data : TPM2B_MAX_BUFFER to unsigned char* */
+    memcpy(out, out_data->buffer, out_data->size);
+    out[out_data->size] = '\0';
+
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    return out_data->size;
+
+error :
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    if (tpm2Data)
+        OPENSSL_free(tpm2Data);
+    return 0;
+}
+
+static int
+tpm2_cipher_cleanup(EVP_CIPHER_CTX *ctx)
+{
+    (void)(ctx);
+    DBG("Cleaning up ...\n");
+    return 1;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+static EVP_CIPHER tpm2_aes_256_ofb =
```

Comment:
Sorry for being so ignorant, but could you explain the 256 and ofb128 to me ?
I don't know what they mean.
Does that mean that aes 128 is not supported ?

I guess a comment in code would also be good...

Reply:
So this is about the **selector function** and **NIDS**, I guess. So will try to explain my understanding of it.

So at first, you can try to type a `openssl enc -ciphers` on your machine. You will see there the different openssl symmetric encryption available. We have to define which algos of this list , our engine will implement. Currently, I have defined these two : 
```c
static int tpm2_cipher_nids[] = {
    NID_aes_256_cfb128,
    NID_aes_256_ofb128,
    0
};
```
Which are defined by AES algorithm, 256 bits key size and cfb/ofb encryption mode.
When someone call `openssl enc -aes-256-cfb -e -engine tpm2tss ...`, our engine implementation will be used.

Moreover, IDs are defined [here](https://github.com/openssl/openssl/blob/master/include/openssl/obj_mac.h#L2735) in the openssl repo. 

As i said before I took the gost engine as example, we can see [here](https://github.com/gost-engine/engine/blob/master/gost_eng.c#L345) the implementation of its cipher selector. Also another really simple cipher engine exits [here](https://github.com/bigbrett/wsaesengine/blob/master/src/wsaesengine.c#L193).

Reply:
If you have other question, don't hesistate. I will try to be the clearest possible and explain my understanding.